### ISSUE TITLE
vote: Create a repository for Notation GitHub Actions

### ISSUE THREAD

**Issue by FeynmanZhou** [notaryproject/.github#30]
Hi @notaryproject/notaryproject-governance-maintainers ,
As we had discussed in https://github.com/notaryproject/notation/issues/544 and along with more users' requests on Notation GitHub Actions, I would like to apply for creating a new repository for Notation GitHub Actions. We plan to create the first version of Notation GitHub Actions after the Notation v1.0.0 release.  It would be easier for users to set up Notation to their GitHub CI/CD workflows.

@Duffney is working on a prototype for Notation GitHub Actions and @Miranlfk also shared his interest in contributing to this project in the community. 

@notaryproject/notaryproject-governance-maintainers If you agree with creating a new repository for the Notation GitHub Actions project, please comment below. Thanks!

**Comment by toddysm** [comment]
LGTM

**Comment by priteshbandi** [comment]
LGTM

@FeynmanZhou Do you have a suggestion for repo's name?

**Comment by yizha1** [comment]
LGTM

**Comment by iamsamirzon** [comment]
LGTM - But I will want us to ensure we support both sign and verify action

**Comment by Duffney** [comment]
> LGTM
> 
> @FeynmanZhou Do you have a suggestion for repo's name?

My suggestion is to create two actions: one for installing notation on the runner and another for signing. For names I'd suggest `setup_notation` as the GH action that installs the CLI because it follows GH action naming norms. For the other GH action responsible for signing, I'd like to suggest `notation-sign-action` as the name. And each action should have its own repository.

**Comment by iamsamirzon** [comment]
> 

@Duffney  - For both the setup and sign action, would you keep the "Plugin" and "Key" as arguments that can be passed. This way, customers can use it with the plugin of their choices. Thoughts? 

Also, are you planning to create a Github action to verify the signature too? It will require "Plugin" argument as well

**Comment by SteveLasker** [comment]
LGTM

**Comment by FeynmanZhou** [comment]
> > LGTM
> > @FeynmanZhou Do you have a suggestion for repo's name?
> 
> My suggestion is to create two actions: one for installing notation on the runner and another for signing. For names I'd suggest `setup-notation` as the GH action that installs the CLI because it follows GH action naming norms. For the other GH action responsible for signing, I'd like to suggest `notation-sign-action` as the name. And each action should have its own repository.

@Duffney  How about `notation-verify-action`? Does it require another repo to host the verification actions?

**Comment by FeynmanZhou** [comment]
Per the discussion in the [Notary community meeting on June 12](https://hackmd.io/_vrqBGAOSUC_VWvFzWruZw?view#Notes), we have the consensus that the first version of GitHub Actions will focus on Notation v1.0.0 installation only. Another reason is that the single actions can be published to GitHub Marketplace without the GitHub review process. Naming this GitHub Actions `setup-notation` sounds good to me.

For `sign` and `verify` actions, we may need to figure out how to securely store and configure the signing key and certificate. This can be designed and implemented in the next iteration.

**Comment by Duffney** [comment]
> > 
> 
> @Duffney - For both the setup and sign action, would you keep the "Plugin" and "Key" as arguments that can be passed. This way, customers can use it with the plugin of their choices. Thoughts?
> 
> Also, are you planning to create a Github action to verify the signature too? It will require "Plugin" argument as well

@iamsamirzon, my thoughts are to keep those args so that it can configure the plugin as well as install the notation binary for the `setup_notation` action. If the user had to configure the plugin separately, I think that would worsen the developer experience. 

As for the sign and verify steps, I haven't put much thought into what they'd look like yet. A common pattern is to handle the setup and let the user define how to use the CLI with Bash in the workflow. That said, if a sign \ verify action were developed, they'd likely need to also include the key and plugin args.

**Comment by yizha1** [comment]
We have `6` out `9` maintainers LGTM, I think we are ready to go to create sub-repo named `setup_notation`. Sign and verify actions will come in later phase. cc:// @FeynmanZhou

**Comment by toddysm** [comment]
Let's hold on creating this repo after the 1.0.0 release.

**Comment by shizhMSFT** [comment]
@yizha1 The repo name should be `setup-notation` instead of `setup_notation`.

**Comment by yizha1** [comment]
Thanks @shizhMSFT . Sorry, my fault. The repo name should be `setup-notation` as you pointed out.

**Comment by toddysm** [comment]
I am about to create the repo but the name is a bit confusing. If the repo will contain only the GHA code, can we prefix the it with `gha` to make sure it is clear what the repo is for? The full name of the repo will be `gha-setup-notation`. This will differentiate if rom other CI/CD tasks like for example `gitlab-setup-notation`, `ado-setup-notation`, `circleci-setup-notation` and so on. Unless we are saying that all `setup-notation` code for all CI/CD systems will be in the same repo

**Comment by toddysm** [comment]
Revising my proposal above after long discussion on Slack. I would propose to name it `setup-notation-action` or `notation-action` and have it look like this in a workflow:

```
steps:
  - uses: notaryproject/setup-notation-action@v1
    with:
      version: 1.0.0
  - run: notation version
```

Or:
```
steps:
  - uses: notaryproject/notation-action@v1
    with:
      version: 1.0.0
  - run: notation version
```

**Comment by sajayantony** [comment]
I think this is far and @toddysm also called out another example of https://github.com/aquasecurity/trivy-action
The non-action repos showed that they were under the actions org. and hence did have some expectation of it being an action.

**Comment by sajayantony** [comment]
`setup-notation-action` SGTM. 
non-maintainer vote.

**Comment by yizha1** [comment]
One comment. If we introduce notation-signing-actions in the near future, will it be a new repository that we need to vote again? IMO, this needs to be clarified and aligned.  /cc @FeynmanZhou This could be a topic for the next community meeting.

**Comment by shizhMSFT** [comment]
This is the doc I have followed for oras: https://docs.github.com/en/actions/creating-actions/developing-a-third-party-cli-action

It states
> This type of action (an action that sets up a tool) is often named `setup-$TOOL`.

It also mentioned some well-known actions:
- https://github.com/ruby/setup-ruby
- https://github.com/google-github-actions/setup-gcloud
- https://github.com/hashicorp/setup-terraform

**Comment by FeynmanZhou** [comment]
> I am about to create the repo but the name is a bit confusing. If the repo will contain only the GHA code, can we prefix the it with gha to make sure it is clear what the repo is for? The full name of the repo will be gha-setup-notation. This will differentiate if rom other CI/CD tasks like for example gitlab-setup-notation, ado-setup-notation, circleci-setup-notation and so on. Unless we are saying that all setup-notation code for all CI/CD systems will be in the same repo.

We will not place all setup-notation code in the CI/CD systems code since this is not a good maintenance practice. If we want to keep the repo name simple but straightforward enough, `notation-action` is a good option for the repository name. We can explain this repo by adding a description on its About. This name is also reusable if we change its scope from single setup action to multi-actions (setup, sign, verify) in the same repository.

**Comment by shizhMSFT** [comment]
An alternative option is a repo with multiple actions although it requires GitHub review for [pulishing](https://docs.github.com/en/actions/creating-actions/publishing-actions-in-github-marketplace#about-publishing-actions).

In this alternative option, we can create a single repo named `notation-actions`. Under this repo, we can create multiple actions, such as
- `notaryproject/notation-actions/setup`
- `notaryproject/notation-actions/sign`
- `notaryproject/notation-actions/verify`

Example workflow file can be:
```yaml
name: release-ghcr

on:
  push:
    tags:
      - v*

jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: docker login
        uses: docker/login-action@v2
        with:
          registry: ghcr.io
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
      - name: docker build
        run: |          
          docker buildx create --use
          docker buildx build -t ${{ steps.prepare.outputs.ref }} --push .
      - name: setup notation # optional for default version of the notation CLI
        uses: notaryproject/notation-actions/setup
        with:
          version: v1.0.0
      - name: sign released image # demo for signing
        uses: notaryproject/notation-actions/sign
        with:
          target-ref: ${{ steps.prepare.outputs.ref }}
          signature-format: cose
          plugin-name: example-signer # fake plugin for demonstration purpose only
          plugin-ref: github.com/shizhMSFT/example-signer@sha256:864fa414b2b5df4709bdcd95d10ff65d98445d4418e8783a313b4a318d454498
          plugin-config: foo=bar
          key-id: 6afc3425-96af-405f-aad7-55de240115a0
      - name: verify release image # demo for verification
        uses: notaryproject/notation-actions/verify
        with:
          target-ref: ${{ steps.prepare.outputs.ref }}
          trust-policy: .github/notation/trustpolicy.json
          trust-store: .github/notation/truststore/
      - name: clear
        if: always()
        run: |
          rm -f ${HOME}/.docker/config.json
```

/cc @toddysm @sajayantony @FeynmanZhou @Two-Hearts

**Comment by shizhMSFT** [comment]
> Is there an example for other project that took this approach?

@toddysm Here are some examples:
- https://github.com/github/codeql-action
  - `github/codeql-action/init`
  - `github/codeql-action/analyze`
- https://github.com/apache/skywalking-eyes
  - `apache/skywalking-eyes/header`
  - `apache/skywalking-eyes/dependency`

**Comment by toddysm** [comment]
This sounds good to me. I will go ahead and will create a `notation-action` repo for all the actions that we want to have.

**Comment by sajayantony** [comment]
+1. Really  like the codeql model.

**Comment by toddysm** [comment]
Created the repository. Need approvals of the following PR in the repo: https://github.com/notaryproject/notation-action/pull/1 before we can proceed with the creation of the GH team and giving rights to folks to start coding.
